### ISSUE TITLE
Support GitHub merge queues when RTD PR builds are required to merge

### ISSUE THREAD

**Issue by gipert** [readthedocs/readthedocs.org#10021]
## Details

I tried to enable [GitHub's merge queue](https://github.blog/changelog/2023-02-08-pull-request-merge-queue-public-beta) on a project that enables [reporting PR builds on GitHub](https://docs.readthedocs.io/en/stable/pull-requests.html) and sets it as a [required status check before merging](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#require-status-checks-before-merging).

## Expected Result

The PR is automatically merged after all checks have passed.

## Actual Result

GitHub waits for ReadTheDocs to report the PR status check, which is never reported. The PR stays in the queue and is never merged.

## Related?

I had to add `on: merge_group:` [[ref]](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue#triggering-merge-group-checks-with-github-actions) to my GHA workflow file to fix a similar problem with the other required status checks run with GitHub Actions.

**Comment by webknjaz** [comment]
@humitos I was just about to file a similar report. This is a blocker for projects wanting to use RTD with Merge Queues enabled.

I want to add that GitHub issues a `push` event in addition to `merge_group`, but RTD doesn't seem to react to it either.

This is what a deadlock this causes looks like:
![Screenshot_2023-10-26-19-21-01-12_40deb401b9ffe8e1df2f1cc5ba480b12](https://github.com/readthedocs/readthedocs.org/assets/578543/7b0e0370-d6df-4a7f-a454-3e43cd54b862)

**Comment by jeff-m-sullivan** [comment]
Having the same issue. pre-commit.ci is the only other third part CI service I'm currently using, and it worked without any configuration changes.

**Comment by humitos** [comment]
@stsewd have you used merge queues? I'm not sure how it works internally, maybe we need to listen to a different event to re-run the build?

> I want to add that GitHub issues a `push` event in addition to `merge_group`

@webknjaz do you a link to the docs where this is explained?

**Comment by jeff-m-sullivan** [comment]
@humitos I think this is the relevant documentation: https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue#how-merge-queues-work

**Comment by stsewd** [comment]
> @stsewd have you used merge queues? I'm not sure how it works internally, maybe we need to listen to a different event to re-run the build?

Haven't used them. But I think the problem is that we report the commit status only on open PRs. Merge queues create a separate branch, but it doesn't create a PR. Should be fixed once we implement https://github.com/readthedocs/readthedocs.org/issues/11739.

**Comment by webknjaz** [comment]
@humitos I think the note in this section https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue#triggering-merge-group-checks-with-github-actions is a more accurate reference. It's kinda obvious that since it makes temporary branches, that action by itself triggers a `push` event just like any other branch creations. It's important to understand that those are ephemeral and are deleted quite soon, once the related PRs are kicked from the queue for various reasons (could be a failure of a check required in branch protection or success of all required checks meaning that the target branch could be fast-forwarded to that commit; could also be a user triggered termination).

@stsewd aren't there also statuses for branches? FWIW, the problem to solve here is the slug to be used in deployment URLs for such ephemeral builds. They won't follow the PR schema because they may correspond to multiple PRs merged-in at the same time, with the target branch not yet fast-forwarded to that state.

**Comment by webknjaz** [comment]
Note that the “two events” thing can cause double builds if you don't deduplicate them. Here's what I tend to do in GHA to run on `merge_queue`, but not `push`: https://github.com/ansible/awx-plugins/blob/0d569b5/.github/workflows/ci-cd.yml#L10.

**Comment by webknjaz** [comment]
> Should be fixed once we implement [#11739](https://github.com/readthedocs/readthedocs.org/issues/11739).

Yes, it sounds like this would fix the problem.

**Comment by stsewd** [comment]
> @stsewd aren't there also statuses for branches?

No, we just report statuses for commits from open PRs.

> FWIW, the problem to solve here is the slug to be used in deployment URLs for such ephemeral builds.

Haven't looked, but if they are normal branches, RTD should just create a new build (users may need to create a rule to activate and build those branches automatically on RTD).

**Comment by webknjaz** [comment]
> Haven't looked, but if they are normal branches, RTD should just create a new build

Yes, they are regular Git branches having more than one slash in their names.

**Comment by rashidnhm** [comment]
Hello, wondering if there is any updates on this, or if there is a way to help with the development of this feature. We wanted to use merge queues but require RTD builds on PRs. 

One solution was to ensure the build on RTD passes before adding a PR to the queue. However, this isn't ideal as it's not clear if a previous merge to main branch could have cause RTD build failures.

**Comment by stsewd** [comment]
Hi @rashidnhm, no updates so far. We may work on this once we have migrated our GH OAuth app to a GitHub App (https://github.com/readthedocs/readthedocs.org/pull/11942).

**Comment by trooz** [comment]
I'm facing this issue too.

**Comment by ssbarnea** [comment]
Any update on this? It forces us to drop RTD builds from being required before merge.

I suspect that one we could attempt to use the newer rulesets to split the branch protections in two groups: those that are merge_groups and the normal ones. Still, this makes the maintenance of the security settings to complex as it would be too each to make a change in one and not in the other. So, the only alterntive is not remove RTD build from being a required check.

**Comment by humitos** [comment]
We have done the work to migrate to the new GitHub App, but we haven't enabled it to everything yet. We were testing it internally. We plan to open the beta in the following weeks.

I'm not 100% sure if that will immediately add support for merge queues or if there is something else we need to add to our workflow cc @stsewd

**Comment by webknjaz** [comment]
@humitos I would imagine supporting merge queues is orthogonal to GH Apps. The only thing that they'll let you do in terms of statuses, would be using Checks API and not just the old Statuses API. However, from the earlier discussions, I'm under the impression that the problem is that the builds aren't getting triggered for the respective event.

**Comment by stsewd** [comment]
> I'm not 100% sure if that will immediately add support for merge queues or if there is something else we need to add to our workflow cc @stsewd 

There is no need for the GH app, we just need to start sending build statuses for all commits, and trigger builds for the branches created by merge queues, as these branches are ephemeral, we need a way to activate them and delete them automatically (maybe automation rules can help here).

**Comment by humitos** [comment]
> trigger builds for the branches created by merge queues, as these branches are ephemeral, we need a way to activate them and delete them automatically (maybe automation rules can help here).

Hrm, this workflow kind of break our current PR preview behavior then because we will:

- create a `Version` for the PR just opened and trigger a new build for it
- create _another_ `Version` for the ephemeral branch and trigger a build for it

I understand those two versions use the exact same commit to build, so we will be building the same set of docs twice per commit/push, right?

Can't we build once and send the commit status twice? One for the original branch and another for the ephemeral branch?

Also, enabling and activating these ephemeral branches/versions will be pretty noisy in the UI of the dashboard because we will end up with a bunch of temporal versions listed there. I think we will need to think more deeply in a integrated solution here instead of a simple workaround.

**Comment by stsewd** [comment]
> I understand those two versions use the exact same commit to build, so we will be building the same set of docs twice per commit/push, right?

Nope, they are different builds, they have changes from different PRs merged, we can't re-use the result from one build.

> Also, enabling and activating these ephemeral branches/versions will be pretty noisy in the UI of the dashboard because we will end up with a bunch of temporal versions listed there. 

These branches will be activated as hidden versions.

> I think we will need to think more deeply in a integrated solution here instead of a simple workaround.

Not sure if adding more complexity is worth it? GH already threats these as normal branches.

**Comment by humitos** [comment]
> These branches will be activated as hidden versions.

We still show hidden versions in the listing: https://app.readthedocs.org/projects/test-builds/versions/

I suppose we could create/enable these versions as `type=external`, however, and treat them as pull requests --which have special behavior in all these listings and other similar places.

**Comment by stsewd** [comment]
That will break some other code that assumes that external versions are directly attached to a PR.
